### PR TITLE
Compact output mode

### PR DESCRIPTION
**Pull request by guybedford**
This implements a compact output mode for Rollup wrappers and injections.

It doesn't remove whitespace in user code, only for the wrappers etc in the Rollup generation itself.

It will be interesting to see where this goes - perhaps whitespace removal in Rollup itself would be useful, perhaps it won't be great for performance, it's hard to tell.

But this gives some starting point that when combined with translate functions doing per-module minification gives an overall minified output that looks pretty good.

### CODE REVIEW

File: test/form/samples/compact/_expected/iife.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+varfoo=(function(x){'use strict';x=x&&x.hasOwnProperty('default')?x['default']:x;function foo () {
```
Comment (wa-Nadoo) [inline test/form/samples/compact/_expected/iife.js]: varfoo seems as a bug

Comment (wa-Nadoo) [inline test/form/samples/compact/_expected/iife.js]: varfoo seems as a bug

Comment (guybedford) [reply test/form/samples/compact/_expected/iife.js]: Thanks!

File: src/ast/variables/NamespaceVariable.ts [hunk 2]
```diff
@@ -61,27 +64,28 @@ export default class NamespaceVariable extends Variable {
...
 
 		if (options.namespaceToStringTag) {
-			output += `\nif (typeof Symbol !== 'undefined' && Symbol.toStringTag)
-${options.indent}Object.defineProperty(${name}, Symbol.toStringTag, { value: 'Module' });
-else
-${
-				options.indent
-			}Object.defineProperty(${name}, 'toString', { value: function () { return '[object Module]' } });
-${callee}(${name});`;
+			const t = options.indent;
```
Comment (lukastaegert) [inline src/ast/variables/NamespaceVariable.ts]: Test?

Comment (lukastaegert) [reply src/ast/variables/NamespaceVariable.ts]: BTW is there a reason why this is not merged with the previous if statement?

File: src/Chunk.ts [hunk 3]
```diff
@@ -1033,8 +1042,8 @@ export default class Chunk {
 			},
 			options
 		);
-		if (addons.banner) magicString.prepend(addons.banner + '\n');
-		if (addons.footer) magicString.append('\n' + addons.footer);
+		if (addons.banner) magicString.prepend(addons.banner + nl);
+		if (addons.footer) magicString.append(nl + addons.footer);
```
[the remark this answers is not in the document]
Comment (guybedford) [reply src/Chunk.ts]: Good catch.

**Comment by guybedford** [comment]
Thanks for the review, I've updated the PR.

File: src/Chunk.ts [hunk 4]
```diff
@@ -402,7 +402,9 @@ export default class Chunk {
 		}
 	}
 
-	private prepareDynamicImports({ format }: OutputOptions) {
+	private prepareDynamicImports({ format, compact }: OutputOptions) {
+		const _ = compact ? '' : ' ';
+		const n = compact ? '' : '\n';
```
Comment (lukastaegert) [inline src/Chunk.ts]: This one is actually unused

File: test/form/samples/compact/_expected/amd.js [hunk 5]
```diff
@@ -0,0 +1,4 @@
+define(['external'],function(x){'use strict';x=x&&x.hasOwnProperty('default')?x['default']:x;var self = {get default(){returnfoo$$1}};if(typeof Symbol!=='undefined'&&Symbol.toStringTag)Object.defineProperty(self,Symbol.toStringTag,{value:'Module'});elseObject.defineProperty(self,'toString',{value:function(){return'[object Module]';}});/*#__PURE__*/Object.freeze(self);console.log(self);
```
Comment (lukastaegert) [inline test/form/samples/compact/_expected/amd.js]: Now this is why I think we need functional tests and should not stop writing them! There are at least two things wrong here that a functional test could probably have found but a human easily overlooks:
* space missing in `returnfoo$$1`
* space missing in `elseObject`

File: test/form/samples/compact/_expected/amd.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+function foo$$1 () {
+	console.log( x );
+}return foo$$1;});
```
Comment (lukastaegert) [inline test/form/samples/compact/_expected/amd.js]: Also it would be nice to separate the user code from the wrapper via new-lines. Not only is it more consistent with the concept of only compactifying the auto-generated code, it will also break down if the user's code ends with a line comment (would not hurt to add this to the test as well)

File: src/utils/addons.ts [hunk 7]
```diff
@@ -14,14 +14,17 @@ export interface Addons {
 
 export function createAddons(graph: Graph, options: OutputOptions): Promise<Addons> {
 	return Promise.all([
-		collectAddon(graph, options.banner, 'banner'),
-		collectAddon(graph, options.footer, 'footer'),
+		collectAddon(graph, options.banner, 'banner', '\n'),
+		collectAddon(graph, options.footer, 'footer', '\n'),
```
Comment (lukastaegert) [inline src/utils/addons.ts]: This looks much safer 👍

File: test/form/samples/compact/main.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import x from 'external';
+import * as self from './main.js';
```
Comment (lukastaegert) [inline test/form/samples/compact/main.js]: Nice adding this!

File: test/form/samples/compact/_expected/amd.js (hunk 5 above)
Comment (guybedford) [reply test/form/samples/compact/_expected/amd.js]: Well caught - will add these.

File: test/form/samples/compact/_expected/amd.js (hunk 6 above)
Comment (guybedford) [reply test/form/samples/compact/_expected/amd.js]: Good point. Let me rather add the ability to detect and remove trailing line comments.

The intention here is that a plugin can perform minification in the transform hook to get an overall minification build running over per-module minify workers.

**Comment by guybedford** [comment]
Thanks again for the thorough review here and catching the edge cases. I've addressed the changes and included the functional tests you mentioned.

For checking a last line comment on a source I created https://github.com/Rich-Harris/magic-string/pull/142 and added that check here, so that should be merged first.

File: test/form/samples/compact/_expected/amd.js (hunk 6 above)
Comment (lukastaegert) [reply test/form/samples/compact/_expected/amd.js]: But trailing line-comments could also be a purposefully placed footer comment. It might be ok to remove it in some situations but we actually have logic to NOT do this as it is still one goal for rollup to preserve all comments a user added as long as we think they do not describe removed code. If a minifier is doing its thing, it will remove those comments anyway so the presence of such a comment would signal that minification was NOT performed and is NOT the intended goal here.

And line comments are just an example. If you throw a line such as `const x = 3 return foo;` into acorn, it will tell you it is invalid syntax. So you also need to make sure there is a semicolon. Etc. etc.

All in all I am not happy about this approach and would prefer the line-break.

Comment (lukastaegert) [reply test/form/samples/compact/_expected/amd.js]: On the other hand, there is no reason to keep the semicolon after `return foo$$1`

Comment (lukastaegert) [reply test/form/samples/compact/_expected/amd.js]: Oh, reading the actual code you wrote, you actually DO preserve the trailing comments. So I am cool with that. Though I am not sure we should do it

Nevertheless, we should also detect for the situation where the last line is non-empty, does not contain a line-comment and where the last non-whitespace character is not a semicolon in which cases I would suggest to append a semicolon instead.

File: test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+define(['exports'],function(exports){'use strict';function fn () {
+  console.log('lib2 fn');
+}function fn$1 () {
```
Comment (lukastaegert) [inline test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js]: This only works because function declarations do not require semicolons to be on the same line as other statements. The same does not hold for e.g. other declarations—
```javascript
const x =3function y(){}
```
would NOT be valid code, even if you add a space in between. A semicolon would be the proper solution here.

File: test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+  fn();
+  console.log('dep2 fn');
+}exports.a=fn$1;});
```
Comment (lukastaegert) [inline test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js]: and again we should consider adding semicolons even if it means adding empty statements in some situations. Alternatively, line-breaks might work as well but would not solve issues where the succeeding code starts with an opening bracket.

File: test/function/samples/compact/_config.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+module.exports = {
```
Comment (lukastaegert) [inline test/function/samples/compact/_config.js]: 👍 for adding this test

File: test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js (hunk 9 above)
Comment (guybedford) [reply test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js]: Note that in the `const x = 3` case, the VariableDeclaration render function will ensure a trailing semicolon already for us.

**Comment by guybedford** [comment]
I've added a fix here for #2110.

I think we shouldn't add a semicolon for the compact case actually - since these cases are already handled by the renderer itself in the individual node renders, which seems exactly how this should be handled to me.

File: test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js (hunk 9 above)
Comment (lukastaegert) [reply test/chunking-form/samples/chunking-compact/_expected/amd/chunk-1e9855f7.js]: Oh, you are right! For declarations and expression statements, this is taken care of 👍

**Comment by guybedford** [comment]
With the magic string update, the tests are now passing here.

@lukastaegert happy to keep this experimental for a while, and just let me know if you want to rename to `experimentalCompact` for example.

**Comment by tivac** [comment]
What's the advantage to having rollup do this itself (though only for the rollup-added wrapping code) as opposed to doing whole-app compaction once the bundle is generated?

**Comment by guybedford** [comment]
@tivac it's an experimental step along the path to turning Rollup into a minifier. For example, mangling is an easy feature for Rollup to add, which when combined with this gets us there. I'm then experimenting with per-source minification for a fast build workflow. Has to be weighed up against whole-graph minification optimizations, but so far there doesn't seem to be much loss. 2/3 build time feels like a win to me so far, but as I say it's experimental.

**Comment by lukastaegert** [comment]
@guybedford I do not think we need an `experimental` option for this. Even though the exact results of this option may change, I would expect it to remain valid at least for a while.
